### ISSUE TITLE
nn_base missing stage5 conv_block

### ISSUE THREAD

**Issue by thealchemist89** [yhenon/keras-frcnn#122]
Hi @yhenon ,

I was looking through the network in resnet.py under the function nn_base(). I notice that the function is missing stage5 conv_block and identity_block to take into account of filter [512, 512, 2048].

Is there a reason why this was taken out when compared to the default Keras resnet50 architecture?

**Comment by thealchemist89** [comment]
Okay, I'm closing this. I suppose the reason for this is to adhere to the Faster RCNN paper on 256-d feature vector for bbox coordinates prediction and classification of object.

**Comment by thealchemist89** [comment]
Sorry for re-opening this ticket. I notice that in resnet.py, under the function rpn(base_layers,num_anchors), I observed that the convolution layer is defined to be a 512-d output. 

Q1. Is this suppose to be 256-d as suggested by the paper for ZF net? Or are we following 512-d as per VGG architecture even though we are using ResNet50?

Q2. Referring to my earlier question then, should we be taking into account of the Stage 5 conv_block that is apparently missing in resnet.py under the function nn_base()?

I would appreciate if anyone out there who is experienced with deep learning could shine some light into this. Thanks everyone for reading :)

**Comment by jahdiel** [comment]
This dimensions are supposed to be chosen based on the fully connected layers of the original architecture.  Therefore I suspect 512-d is not the correct shape for Resnet-50, might be wrong. @yhenon is there a specific reason for 512-d?

**Comment by yhenon** [comment]
I've looked over this and I'm pretty sure that is the correct size for that layer. My implementation matches what I found here (https://github.com/rbgirshick/py-faster-rcnn/issues/62). Also, 512 is the size of the layer 5a that follows the layer 4f. 
The general implementation does match what is described in the resnet paper (although with resnet-5, not resnet-101). The RPN is after stage 4, then the classifier uses stage 5:
>  We compute the full-image shared conv feature maps using those layers
whose strides on the image are no greater than 16 pixels
(i.e., conv1, conv2 x, conv3 x, and conv4 x, totally 91 conv
layers in ResNet-101; Table 1). We consider these layers as
analogous to the 13 conv layers in VGG-16, and by doing
so, both ResNet and VGG-16 have conv feature maps of the
same total stride (16 pixels). These layers are shared by a
region proposal network (RPN, generating 300 proposals)
[32] and a Fast R-CNN detection network [7]. RoI pooling
[7] is performed before conv5 1. On this RoI-pooled
feature, all layers of conv5 x and up are adopted for each
region, playing the roles of VGG-16’s fc layers.

**Comment by jahdiel** [comment]
@yhenon Does this implementation of faster r-cnn use 300 region proposals? And where can I change the amount of region proposals in the code? Thanks.

**Comment by yhenon** [comment]
At train time it uses 300 (see `train_frcnn.py`), `R =
roi_helpers.rpn_to_roi(P_rpn[0], P_rpn[1], C, K.image_dim_ordering(),
use_regr=True, overlap_thresh=0.7, max_boxes=300)`
Not a bad idea to make it a parameter in config.py

**Comment by jahdiel** [comment]
Yes, thanks so much. In the paper they use 2000 region proposals for training and 300 for testing, it should be a good idea to notify this minor change given it can have a substantial influence in the performance of the network.

**Comment by thealchemist89** [comment]
@yhenon , thank you for your reply, much appreciation! If the anchor boxes and scale is reduced, will the detector be able to pick up smaller objects in your experience?

Much appreciation for open sourcing this repo.

**Comment by liuwei16** [comment]
@yhenon I have a confusion about roi_pooling size in resnet.py, why is the pooling size in tensorflow is 14 while 7 in theano?
Is there some specific consideration?
As I know, the default pooing size in original faster rcnn is 7.
Thank you very much.

**Comment by Tveek** [comment]
@yhenon which paper for 
**We compute the full-image shared conv feature maps using those layers
whose strides on the image are no greater than 16 pixels
(i.e., conv1, conv2 x, conv3 x, and conv4 x, totally 91 conv
layers in ResNet-101; Table 1). We consider these layers as
analogous to the 13 conv layers in VGG-16, and by doing
so, both ResNet and VGG-16 have conv feature maps of the
same total stride (16 pixels). These layers are shared by a
region proposal network (RPN, generating 300 proposals)
[32] and a Fast R-CNN detection network [7]. RoI pooling
[7] is performed before conv5 1. On this RoI-pooled
feature, all layers of conv5 x and up are adopted for each
region, playing the roles of VGG-16’s fc layers.**
I can not find this in KaimingHe's 'Deep Residual Learning for Image Recognition
'
